File: code/3-Longest Substring Without Repeating Characters/3-Longest Substring Without Repeating Characters.py
"""
LeetCode 3 - Longest Substring Without Repeating Characters

解题步骤（自动整理）
1. 先把题目目标拆成可操作的子任务（比较/统计/构造等）
2. 选择合适的数据结构并按一次遍历或分治步骤实现核心逻辑
3. 补齐边界条件（空输入、单元素、重复元素等）并返回结果
"""
# Given a string, find the length of the longest substring without repeating characters.

# Examples:

# Given "abcabcbb", the answer is "abc", which the length is 3.
# Given "bbbbb", the answer is "b", with the length of 1.
# Given "pwwkew", the answer is "wke", with the length of 3. Note that the answer must be a substring, "pwke" is a subsequence and not a substring.

# 用hash表，也就是python中的字典dict来存储字符串中的内容

class Solution(object):
    def lengthOfLongestSubstring(self, s):
        """
        :type s: str
        :rtype: int
        """
        dic, start, res = {}, 0, 0
        for i, ch in enumerate(s):
            # 这个enumerate用的很巧妙，精简了很多代码，不然要写成dict[s[i]]
            if ch in dic:
                res = max(res, i-start)
                start = max(start, dic[ch] + 1)
            dic[ch] = i
        return max(res, len(s)-start)
        # 小心没重复的substring


class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        mp = set()
        n = len(s)
        start, res = -1, 0
        for i in range(n):
            if  i != 0:
                mp.remove(s[i - 1])
            while start + 1 < n and  s[start + 1] not in mp:
                mp.add(s[start + 1])
                start += 1
            res = max(res, start - i + 1)
        return res

File: code/3-Longest Substring Without Repeating Characters/3-Longest Substring Without Repeating Characters.cpp
/*Longest Substring Without Repeating Characters
Given a string, find the length of the longest substring without repeating characters. For example, the longest substring without repeating letters for "abcabcbb" is "abc", which the length is 3. For "bbbbb" the longest substring is "b", with the length of 1.

选一个hash表，其实就是通过字符表达的值作为index能找到其在string中位置的储存结构，然后从左往右扫描一遍，准备一个start指针，凡是找到一个重复的字符，那么先算一下是不是当时情况下的最长，然后更新最长值，接着把之前的start到被重复的字符的hash值重新设置为-1，然后把当前的hash值更新

*/

class Solution {
public:
    int lengthOfLongestSubstring(string s) {
        unordered_set<char> mp;
        int start = -1, res = 0, n = s.size();
        for (int i = 0; i < n; i ++) {
            if (i != 0) {
                mp.erase(s[i - 1]);
            } 
            while (start + 1 < n && !mp.count(s[start + 1])) {
                mp.emplace(s[start + 1]);
                start ++;
            }
            res = max(res, start - i + 1);
        }
        return res;
    }
};